File: src/main/scala/net/machinemuse/powersuits/client/render/modelspec/RenderPart.scala
package net.machinemuse.powersuits.client.render.modelspec

import net.machinemuse.general.NBTTagAccessor
import net.machinemuse.numina.geometry.Colour
import net.machinemuse.numina.render.MuseTextureUtils
import net.machinemuse.powersuits.client.render.item.ArmorModel
import net.machinemuse.utils.render.Render
import net.minecraft.client.model.{ModelBase, ModelRenderer}
import net.minecraft.client.renderer.{OpenGlHelper, RenderHelper}
import net.minecraft.nbt.NBTTagCompound
import org.lwjgl.opengl.GL11

/**
  * Author: MachineMuse (Claire Semple)
  * Created: 4:16 AM, 29/04/13
  */

class RenderPart(base: ModelBase, val parent: ModelRenderer) extends ModelRenderer(base) {
  override def render(scale: Float) {
    val renderSpec = ArmorModel.instance.renderSpec
    import scala.collection.JavaConverters._
    val colours = renderSpec.getIntArray("colours")

    for (nbt <- NBTTagAccessor.getValues(renderSpec).asScala) {
      val part = ModelRegistry.getPart(nbt)
      if (part.isDefined) {
        if (part.get.slot == ArmorModel.instance.visibleSection && part.get.morph.apply(ArmorModel.instance) == parent) {
          var prevBrightX: Float = OpenGlHelper.lastBrightnessX;
          var prevBrightY: Float = OpenGlHelper.lastBrightnessY;

          // GLOW stuff on
          if (part.get.getGlow(nbt)) {
            GL11.glPushAttrib(GL11.GL_LIGHTING_BIT)
            RenderHelper.disableStandardItemLighting()
            OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, 240.0F, 240.0F)
          }

          GL11.glPushMatrix()
          GL11.glScaled(scale, scale, scale)
          MuseTextureUtils.bindTexture(part.get.getTexture(nbt))
          applyTransform()
          val ix = part.get.getColourIndex(nbt)
          if (ix < colours.size && ix >= 0) {
            Colour.doGLByInt(colours(ix))
          }
          part.get.modelSpec.applyOffsetAndRotation() // not yet implemented
          part.get.modelSpec.model.renderPart(part.get.partName)
          Colour.WHITE.doGL()
          GL11.glPopMatrix()

          // GLOW stuff off
          if (part.get.getGlow(nbt)) {
            OpenGlHelper.setLightmapTextureCoords(OpenGlHelper.lightmapTexUnit, prevBrightX, prevBrightY)
            GL11.glPopAttrib()
          }
        }
      }
    }
  }

  val degrad = 180F / Math.PI.asInstanceOf[Float]

//  def withMaybeGlow[A](part: ModelPartSpec, nbt: NBTTagCompound)(r: Render[A]): Render[A] = {
//    if (part.getGlow(nbt)) {
//      Render withGlow r
//    } else {
//      r
//    }
//  }

  def applyTransform() {
    //    glTranslatef(rotationPointX, rotationPointY, rotationPointZ)
    //    glRotatef(rotateAngleZ * degrad, 0.0F, 0.0F, 1.0F)
    //    glRotatef(rotateAngleY * degrad, 0.0F, 1.0F, 0.0F)
    //    glRotatef(rotateAngleX * degrad, 1.0F, 0.0F, 0.0F)
    GL11.glRotatef(180, 1.0F, 0.0F, 0.0F)

    GL11.glTranslated(offsetX, offsetY - 26, offsetZ)

  }
}

File: src/main/java/net/machinemuse/powersuits/client/render/modelspec/RenderPart.java
package net.machinemuse.powersuits.client.render.modelspec;

import net.machinemuse.numina.client.render.RenderState;
import net.machinemuse.numina.client.render.modelspec.ModelPartSpec;
import net.machinemuse.numina.client.render.modelspec.ModelRegistry;
import net.machinemuse.numina.client.render.modelspec.PartSpecBase;
import net.machinemuse.numina.constants.ModelSpecTags;
import net.machinemuse.numina.math.Colour;
import net.machinemuse.powersuits.client.model.item.armor.ArmorModelInstance;
import net.machinemuse.powersuits.client.model.item.armor.IArmorModel;
import net.machinemuse.powersuits.utils.nbt.NBTTagAccessor;
import net.minecraft.client.Minecraft;
import net.minecraft.client.model.ModelBase;
import net.minecraft.client.model.ModelRenderer;
import net.minecraft.client.renderer.BufferBuilder;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.block.model.BakedQuad;
import net.minecraft.client.renderer.texture.TextureMap;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraftforge.client.ForgeHooksClient;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.lwjgl.opengl.GL11;

import java.util.List;

/**
 * Author: MachineMuse (Claire Semple)
 * Created: 4:16 AM, 29/04/13
 * <p>
 * Ported to Java by lehjr on 11/6/16.
 */
@SideOnly(Side.CLIENT)
public class RenderPart extends ModelRenderer {
    ModelRenderer parent;

    public RenderPart(ModelBase base, ModelRenderer parent) {
        super(base);
        this.parent = parent;
    }

    @Override
    public void render(float scale) {
        NBTTagCompound renderSpec = ((IArmorModel) (ArmorModelInstance.getInstance())).getRenderSpec();
        if (renderSpec == null)
            return;

        int[] colours = renderSpec.getIntArray(ModelSpecTags.TAG_COLOURS);
        if (colours.length == 0)
            colours = new int[]{Colour.WHITE.getInt()};

        int partColor;
        for (NBTTagCompound nbt : NBTTagAccessor.getValues(renderSpec)) {
            PartSpecBase part = ModelRegistry.getInstance().getPart(nbt);
            if (part != null && part instanceof ModelPartSpec) {
                if (part.getBinding().getSlot() == ((IArmorModel) (ArmorModelInstance.getInstance())).getVisibleSection()
                        && part.getBinding().getTarget().apply(ArmorModelInstance.getInstance()) == parent) {
                    List<BakedQuad> quadList = ((ModelPartSpec) part).getQuads();
                    if (!quadList.isEmpty()) {
                        int ix = part.getColourIndex(nbt);

                        // checks the range of the index to avoid errors OpenGL or crashing
                        if (ix < colours.length && ix >= 0) partColor = colours[ix];
                        else partColor = Colour.WHITE.getInt();

                        // GLOW stuff on
                        if (((ModelPartSpec) part).getGlow(nbt))
                            RenderState.glowOn();

                        GlStateManager.pushMatrix();
                        GlStateManager.scale(scale, scale, scale);
                        applyTransform();
                        Minecraft.getMinecraft().renderEngine.bindTexture(TextureMap.LOCATION_BLOCKS_TEXTURE);
                        Tessellator tess = Tessellator.getInstance();
                        BufferBuilder buffer = tess.getBuffer();
                        buffer.begin(GL11.GL_QUADS, DefaultVertexFormats.ITEM);

                        for (BakedQuad quad : ((ModelPartSpec) part).getQuads()) {
                            buffer.addVertexData(quad.getVertexData());
                            ForgeHooksClient.putQuadColor(buffer, quad, partColor);
                        }
                        tess.draw();

                        GlStateManager.popMatrix();

                        //Glow stuff off
                        if (((ModelPartSpec) part).getGlow(nbt))
                            RenderState.glowOff();
                    }
                }
            }
        }
    }

    private void applyTransform() {
//        float degrad = (float) (180F / Math.PI);
//        GL11.glTranslatef(rotationPointX, rotationPointY, rotationPointZ);
//        GL11.glRotatef(rotateAngleZ * degrad, 0.0F, 0.0F, 1.0F);
//        GL11.glRotatef(rotateAngleY * degrad, 0.0F, 1.0F, 0.0F);
//        GL11.glRotatef(rotateAngleX * degrad, 1.0F, 0.0F, 0.0F);
        GlStateManager.rotate(180, 1.0F, 0.0F, 0.0F);
        GlStateManager.translate(offsetX, offsetY - 26, offsetZ);
    }
}